lib/color/color.go

```diff
@@ -1,11 +1,74 @@
 package color
 
 import (
+	"fmt"
+	"regexp"
+
 	"github.com/lucasb-eyer/go-colorful"
 	"github.com/mazznoer/csscolorparser"
 )
 
+var themeColorRegex = regexp.MustCompile(`^N[1-7]|B[1-6]|AA[245]|AB[45]$`)
+
+func IsThemeColor(colorString string) bool {
+	return themeColorRegex.Match([]byte(colorString))
+}
+
 func Darken(colorString string) (string, error) {
+	if IsThemeColor(colorString) {
+		switch {
+		case colorString[0] == 'B':
+			switch colorString[1] {
+			case '1', '2':
+				return B1, nil
+			case '3':
+				return B2, nil
+			case '4':
+				return B3, nil
+			case '5':
+				return B4, nil
+			case '6':
+				return B5, nil
+			}
+
+		case colorString[0:2] == "AA":
+			switch colorString[2] {
+			case '2', '4':
+				return AA2, nil
+			case '5':
+				return AA4, nil
+			}
+
+		case colorString[0:2] == "AB":
+			switch colorString[2] {
+			case '4', '5':
+				return AB4, nil
+			}
```

Comment:
@vfosnar i just realized, is there a reason why 4 and 5 are aggregated here?

Reply:
i'm guessing it's a mistake (even though the code here looks deliberate). fixed here: https://github.com/terrastruct/d2/pull/883

Reply:
Well, AFAIR this code should shift any given theme color to the next one and when it was the first one, it will return itself. Thus if we have 2 colors, we return same color in both cases.

This whole shifting code was intended more like a "somewhat working" solution and I thought I left a TODO explaining it somewhere there and that a better solution may be available but would probably require adding all rn missing colors to themes and tweaking them a little.

also I think this method does the opposite in dark mode, it lightens the value, becase dark themes have flipped pallet.

reference to why I thought shifting could work:
https://github.com/terrastruct/d2/tree/master/d2themes#color-coding-example

Reply:
oh shoot yeah i didn't look at the function name, this makes sense. hmmm, will think about this.

Reply:
https://github.com/terrastruct/d2/issues/887